## **Task 3**

In [2]:
import pandas as pd
import numpy as np

In [5]:
# Read the CSV files into Pandas DataFrames
orders_df = pd.read_csv("orders.csv")
order_details_df = pd.read_csv("order_details.csv")
pizzas_df = pd.read_csv("pizzas.csv")
pizza_types_df = pd.read_csv("pizza_types.csv", encoding="latin1")

# Display the first 5 rows of each DataFrame
print("\nFirst 5 rows of 'Orders':")
display(orders_df.head())

print("\nFirst 5 rows of 'Order Details':")
display(order_details_df.head())

print("\nFirst 5 rows of 'Pizzas':")
display(pizzas_df.head())

print("\nFirst 5 rows of 'Pizza Types':")
display(pizza_types_df.head())

# Display columns and their data types for each DataFrame
print("\nColumn information for 'Orders':")
print(orders_df.info())

print("\nColumn information for 'Order Details':")
print(order_details_df.info())

print("\nColumn information for 'Pizzas':")
print(pizzas_df.info())

print("\nColumn information for 'Pizza Types':")
print(pizza_types_df.info())


First 5 rows of 'Orders':


order_id        date      time
0         1  2015-01-01  11:38:36
1         2  2015-01-01  11:57:40
2         3  2015-01-01  12:12:28
3         4  2015-01-01  12:16:31
4         5  2015-01-01  12:21:30


First 5 rows of 'Order Details':


order_details_id  order_id       pizza_id  quantity
0                 1         1     hawaiian_m         1
1                 2         2  classic_dlx_m         1
2                 3         2  five_cheese_l         1
3                 4         2    ital_supr_l         1
4                 5         2     mexicana_m         1


First 5 rows of 'Pizzas':


pizza_id pizza_type_id size  price
0   bbq_ckn_s       bbq_ckn    S  12.75
1   bbq_ckn_m       bbq_ckn    M  16.75
2   bbq_ckn_l       bbq_ckn    L  20.75
3  cali_ckn_s      cali_ckn    S  12.75
4  cali_ckn_m      cali_ckn    M  16.75


First 5 rows of 'Pizza Types':


pizza_type_id                          name category  \
0       bbq_ckn    The Barbecue Chicken Pizza  Chicken   
1      cali_ckn  The California Chicken Pizza  Chicken   
2   ckn_alfredo     The Chicken Alfredo Pizza  Chicken   
3     ckn_pesto       The Chicken Pesto Pizza  Chicken   
4    southw_ckn   The Southwest Chicken Pizza  Chicken   

                                         ingredients  
0  Barbecued Chicken, Red Peppers, Green Peppers,...  
1  Chicken, Artichoke, Spinach, Garlic, Jalapeno ...  
2  Chicken, Red Onions, Red Peppers, Mushrooms, A...  
3  Chicken, Tomatoes, Red Peppers, Spinach, Garli...  
4  Chicken, Tomatoes, Red Peppers, Red Onions, Ja...


Column information for 'Orders':
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21350 entries, 0 to 21349
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   order_id  21350 non-null  int64 
 1   date      21350 non-null  object
 2   time      21350 non-null  object
dtypes: int64(1), object(2)
memory usage: 500.5+ KB
None

Column information for 'Order Details':
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48620 entries, 0 to 48619
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   order_details_id  48620 non-null  int64 
 1   order_id          48620 non-null  int64 
 2   pizza_id          48620 non-null  object
 3   quantity          48620 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 1.5+ MB
None

Column information for 'Pizzas':
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96 entries, 0 to 95
Data columns (total 4 colum

In [6]:
# Merging order_details_df with orders_df on 'order_id'
merged_df = pd.merge(order_details_df, orders_df, on='order_id', how='outer')

# Display the first 5 rows of the merged DataFrame
print("\nFirst 5 rows after merging 'Order Details' and 'Orders':")
display(merged_df.head())

# Merging the result with pizzas_df on 'pizza_id'
merged_df = pd.merge(merged_df, pizzas_df, on='pizza_id', how='outer')

# Display the first 5 rows of the merged DataFrame
print("\nFirst 5 rows after merging with 'Pizzas':")
display(merged_df.head())

# Merging the result with pizza_types_df on 'pizza_type_id'
final_merged_df = pd.merge(merged_df, pizza_types_df, on='pizza_type_id', how='outer')

# Display the first 5 rows of the merged DataFrame
print("\nFirst 5 rows after merging with 'Pizza Types':")
display(final_merged_df.head())

# Display columns and their data types of the merged DataFrame
print("\nColumn information after merging with 'Pizza Types':")
print(final_merged_df.info())


First 5 rows after merging 'Order Details' and 'Orders':


order_details_id  order_id       pizza_id  quantity        date      time
0                 1         1     hawaiian_m         1  2015-01-01  11:38:36
1                 2         2  classic_dlx_m         1  2015-01-01  11:57:40
2                 3         2  five_cheese_l         1  2015-01-01  11:57:40
3                 4         2    ital_supr_l         1  2015-01-01  11:57:40
4                 5         2     mexicana_m         1  2015-01-01  11:57:40


First 5 rows after merging with 'Pizzas':


order_details_id  order_id   pizza_id  quantity        date      time  \
0              26.0      11.0  bbq_ckn_l       1.0  2015-01-01  13:02:59   
1              43.0      17.0  bbq_ckn_l       1.0  2015-01-01  13:53:00   
2              71.0      27.0  bbq_ckn_l       1.0  2015-01-01  15:11:17   
3              83.0      33.0  bbq_ckn_l       1.0  2015-01-01  15:54:08   
4             123.0      51.0  bbq_ckn_l       1.0  2015-01-01  18:48:28   

  pizza_type_id size  price  
0       bbq_ckn    L  20.75  
1       bbq_ckn    L  20.75  
2       bbq_ckn    L  20.75  
3       bbq_ckn    L  20.75  
4       bbq_ckn    L  20.75


First 5 rows after merging with 'Pizza Types':


order_details_id  order_id   pizza_id  quantity        date      time  \
0              26.0      11.0  bbq_ckn_l       1.0  2015-01-01  13:02:59   
1              43.0      17.0  bbq_ckn_l       1.0  2015-01-01  13:53:00   
2              71.0      27.0  bbq_ckn_l       1.0  2015-01-01  15:11:17   
3              83.0      33.0  bbq_ckn_l       1.0  2015-01-01  15:54:08   
4             123.0      51.0  bbq_ckn_l       1.0  2015-01-01  18:48:28   

  pizza_type_id size  price                        name category  \
0       bbq_ckn    L  20.75  The Barbecue Chicken Pizza  Chicken   
1       bbq_ckn    L  20.75  The Barbecue Chicken Pizza  Chicken   
2       bbq_ckn    L  20.75  The Barbecue Chicken Pizza  Chicken   
3       bbq_ckn    L  20.75  The Barbecue Chicken Pizza  Chicken   
4       bbq_ckn    L  20.75  The Barbecue Chicken Pizza  Chicken   

                                         ingredients  
0  Barbecued Chicken, Red Peppers, Green Peppers,...  
1  Barbecued Chicken, Red Peppers, Green Peppers,...  
2  Barbecued Chicken, Red Peppers, Green Peppers,...  
3  Barbecued Chicken, Red Peppers, Green Peppers,...  
4  Barbecued Chicken, Red Peppers, Green Peppers,...


Column information after merging with 'Pizza Types':
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48625 entries, 0 to 48624
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_details_id  48620 non-null  float64
 1   order_id          48620 non-null  float64
 2   pizza_id          48625 non-null  object 
 3   quantity          48620 non-null  float64
 4   date              48620 non-null  object 
 5   time              48620 non-null  object 
 6   pizza_type_id     48625 non-null  object 
 7   size              48625 non-null  object 
 8   price             48625 non-null  float64
 9   name              48625 non-null  object 
 10  category          48625 non-null  object 
 11  ingredients       48625 non-null  object 
dtypes: float64(4), object(8)
memory usage: 4.5+ MB
None


### **Question 1**
1.	What is the total revenue/sales?

In [7]:
# Calculate total revenue by multiplying quantity by price
final_merged_df['revenue'] = final_merged_df['quantity'] * final_merged_df['price']

# Sum the revenue column to get the total revenue
total_revenue = final_merged_df['revenue'].sum()

# Display the total revenue
print(f"The total revenue is: ${total_revenue:,.2f}")

The total revenue is: $817,860.05


### **Question 2**
2.	Find the total quantity sold.

In [9]:
# Calculate the total quantity sold
total_quantity_sold = final_merged_df['quantity'].sum()

# Display the total quantity sold
print(f"The total quantity of pizzas sold is: {total_quantity_sold}")

The total quantity of pizzas sold is: 49574.0


### **Question 3**
3.	Find the total orders.

In [10]:
# Calculate the total number of orders
total_orders = final_merged_df['order_id'].nunique()

# Display the total number of orders
print(f"The total number of orders is: {total_orders:,.0f}")

The total number of orders is: 21,350


### **Question 4**
4.	How many pizza types do they sell?

In [11]:
# Calculate the number of unique pizza types
num_pizza_types = final_merged_df['pizza_type_id'].nunique()

# Display the number of unique pizza types
print(f"The number of unique pizza types sold is: {num_pizza_types}")

The number of unique pizza types sold is: 32


### **Question 5**
5.	Find the average price of the pizzas.

In [12]:
# Calculate the average price of the pizzas
average_price = final_merged_df['price'].mean()

# Display the average price of the pizzas
print(f"The average price of the pizzas is: ${average_price:,.2f}")

The average price of the pizzas is: $16.49


### **Question 6**
6.	What are the peak hours of sales?

In [45]:
# Convert 'time' column to datetime objects with specified format
final_merged_df['time'] = pd.to_datetime(final_merged_df['time'], format='%H:%M:%S')

# Extract hour from the 'time' column
final_merged_df['hour'] = final_merged_df['time'].dt.hour

# Group by hour and count the number of orders
hourly_order_counts = final_merged_df.groupby('hour').size().reset_index(name='order_count')

# Sort by order count to find peak hours
peak_hours = hourly_order_counts.sort_values(by='order_count', ascending=False)

# Display the peak hours
print("Peak hours of sales:\n")
# display(peak_hours)
display(peak_hours.head())

Peak hours of sales:



hour  order_count
3   12.0         6543
4   13.0         6203
9   18.0         5359
8   17.0         5143
10  19.0         4350

The peak hours of sales are between `12:00 PM`and `1:00 PM`.

In [18]:
final_merged_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48625 entries, 0 to 48624
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   order_details_id  48620 non-null  float64       
 1   order_id          48620 non-null  float64       
 2   pizza_id          48625 non-null  object        
 3   quantity          48620 non-null  float64       
 4   date              48620 non-null  object        
 5   time              48620 non-null  datetime64[ns]
 6   pizza_type_id     48625 non-null  object        
 7   size              48625 non-null  object        
 8   price             48625 non-null  float64       
 9   name              48625 non-null  object        
 10  category          48625 non-null  object        
 11  ingredients       48625 non-null  object        
 12  revenue           48620 non-null  float64       
 13  hour              48620 non-null  float64       
dtypes: datetime64[ns](1), 

### **Question 7**
7.	Find the total sales made on each day of the week. Which day of the week is when sales are made the most?

In [21]:
# Convert 'date' column to datetime
final_merged_df['date'] = pd.to_datetime(final_merged_df['date'])

# Extract day of the week
final_merged_df['day_of_week'] = final_merged_df['date'].dt.day_name()

# Group by day of the week and sum the revenue
daily_sales = final_merged_df.groupby('day_of_week')['revenue'].sum().reset_index(name='total_sales')

# Sort by total sales
most_sales_day = daily_sales.sort_values(by='total_sales', ascending=False)

# Display the total sales for each day of the week
print("Total sales for each day of the week:\n")
display(daily_sales)

# Display the day with the most sales
print(f"\n{most_sales_day.iloc[0]['day_of_week']} is the day with the most sales (${most_sales_day.iloc[0]['total_sales']:.2f})")

Total sales for each day of the week:



day_of_week  total_sales
0      Friday    136073.90
1      Monday    107329.55
2    Saturday    123182.40
3      Sunday     99203.50
4    Thursday    123528.50
5     Tuesday    114133.80
6   Wednesday    114408.40


Friday is the day with the most sales ($136073.90)


### **Question 8**
8.	Find the top 5 bestselling pizzas.

In [28]:
# Group by pizza name and sum the quantity
pizza_sales = final_merged_df.groupby('name')['quantity'].sum().reset_index(name='total_quantity')

# Sort by total quantity sold
top_selling_pizzas = pizza_sales.sort_values(by='total_quantity', ascending=False)

# Display the top 5 best-selling pizzas
print("Top 5 best-selling pizzas:\n")
display(top_selling_pizzas.head())

Top 5 best-selling pizzas:



name  total_quantity
7     The Classic Deluxe Pizza          2453.0
0   The Barbecue Chicken Pizza          2432.0
12          The Hawaiian Pizza          2422.0
20         The Pepperoni Pizza          2418.0
30      The Thai Chicken Pizza          2371.0

### **Question 9**
9.	Find the sales made in each month. Any trend noticeable?

In [33]:
import pandas as pd

# Extract month from the 'date' column
final_merged_df['month'] = final_merged_df['date'].dt.month_name()

# Group by month and sum the revenue
monthly_sales = final_merged_df.groupby('month')['revenue'].sum().reset_index()

# Order the months correctly
month_order = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
monthly_sales['month'] = pd.Categorical(monthly_sales['month'], categories=month_order, ordered=True)
monthly_sales = monthly_sales.sort_values('month')

# Display the sales for each month
print("Sales for each month:\n")
display(monthly_sales)

# Analyze trends (basic trend observation)
highest_sales_month = monthly_sales.loc[monthly_sales['revenue'].idxmax(), 'month']
lowest_sales_month = monthly_sales.loc[monthly_sales['revenue'].idxmin(), 'month']

print(f"\nGeneral Trend: Highest sales were in {highest_sales_month}, while lowest sales were in {lowest_sales_month}.")

Sales for each month:



month   revenue
4     January  69793.30
3    February  65159.60
7       March  70397.10
0       April  68736.80
8         May  71402.75
6        June  68230.20
5        July  72557.90
1      August  68278.25
11  September  64180.05
10    October  64027.60
9    November  70395.35
2    December  64701.15


General Trend: Highest sales were in July, while lowest sales were in October.


`Observation`: Sales fluctuate throughout the year, with peaks and troughs.

### **Question 10**
10.	Are there pizza types that are not doing well on the menu?

In [46]:
# Group by pizza type and sum the quantity
pizza_type_sales = final_merged_df.groupby('pizza_type_id')['quantity'].sum().reset_index(name='total_quantity_sold')

# Sort by total quantity sold in ascending order to find the least popular
least_popular_pizza_types = pizza_type_sales.sort_values(by='total_quantity_sold', ascending=True)

# Display the bottom 5 pizza types
print("\nBottom 5 least popular pizza types:\n")
display(least_popular_pizza_types.head())


Bottom 5 least popular pizza types:



pizza_type_id  total_quantity_sold
2     brie_carre                490.0
15  mediterraneo                934.0
3      calabrese                937.0
28  spinach_supr                950.0
23   soppressata                961.0

The pizza type that is not doing well is `brie_carre`. The total quantity sold of `490` is the least